### PR TITLE
Make ChangedElements public

### PR DESCRIPTION
**Pull request by diegopinate**
Make changed elements public and add descriptions so that we can reference the docs in the upcoming APIM tutorials for changed elements API.

### CODE REVIEW

**Review by pmconne — COMMENTED** [review]
Is there some API that accepts or returns ChangedElements? Or is that forthcoming?

**Comment by diegopinate** [comment]
> Is there some API that accepts or returns ChangedElements? Or is that forthcoming?

Outside of iModelJs repo: The Changed Elements REST API returns a JSON object in this format. This API is available now via APIM in PROD.
Inside iModelJs repo: This is resturned in our internal backend vc processing but I don't want to expose as public just yet.

**Comment by pmconne** [comment]
> Inside iModelJs repo: This is returned in our internal backend vc processing but I don't want to expose that as public just yet.

Ok. Ideally after those tutorials are published the ChangedElements docs will link to them.

**Comment by diegopinate** [comment]
> > Inside iModelJs repo: This is returned in our internal backend vc processing but I don't want to expose that as public just yet.
> 
> Ok. Ideally after those tutorials are published the ChangedElements docs will link to them.

Sounds good, I'll keep a note and ensure that's updated.

File: core/common/src/ChangedElements.ts [hunk 1]
```diff
@@ -11,26 +11,57 @@ import { AxisAlignedBox3dProps } from "./geometry/Placement";
...
+   * between the versions being inspected
+   */
   newChecksums?: number[][];
+  /**
+   * Parent ids of the changed elements
+   * Will be "0" if the element has no parent
+   */
   parentIds?: Id64String[];
+  /**
+   * Parent [[ECClass]] ids of the changed elements
```
Comment (pmconne) [inline core/common/src/ChangedElements.ts:62]: This link won't work -`[[SomeThing]]` is short-hand for `[SomeThing]($this-package)` but ECClass is not defined in imodeljs-common.

File: core/common/src/ChangedElements.ts [hunk 2]
```diff
@@ -11,26 +11,57 @@ import { AxisAlignedBox3dProps } from "./geometry/Placement";
  * @module Entities
  */
 
-/** @internal */
+/** Type of change bitflags that show what changed on an element
```
Comment (pmconne) [inline core/common/src/ChangedElements.ts:14]: Can you link to related APIs? e.g.,
`Bitflags describing which aspects of an [Element]($backend) changed as part of a [[ChangedElements]].`
`The [[Placement]] of a [GeometricElement]($backend) changed.`
etc.

**Review by pmconne — CHANGES_REQUESTED** [review]
Couple of nitpicks.

File: core/common/src/ChangedElements.ts [hunk 3]
```diff
@@ -11,26 +11,57 @@ import { AxisAlignedBox3dProps } from "./geometry/Placement";
...
+  /** The placement of a geometric element changed */
   Placement = 0b100,
+  /** Indirect change occurred to this element by a related instance */
   Indirect = 0b1000,
+  /** Hidden properties of the element changed */
   Hidden = 0b10000,
 }
 
-/** @internal */
+/** Changed elements found in a changeset or between a range of changesets
```
Comment (pmconne) [inline core/common/src/ChangedElements.ts:30]: Does `classIds` have the same number of entries as `elements` such that the classId of elements[x] is always classIds[x]?
Is the same true of `modelIds` and `types`?

File: core/common/src/ChangedElements.ts [hunk 4]
```diff
@@ -11,26 +11,57 @@ import { AxisAlignedBox3dProps } from "./geometry/Placement";
...
+  /** Ids of elements that changed */
   elements: Id64String[];
+  /** ECClassIds of elements that changed */
   classIds: Id64String[];
+  /** Operation codes, see [[DbOpcode]] */
   opcodes: number[];
+  /** Type of change bitflags, see [[TypeOfChange]] */
   type: number[];
+  /** Model Ids of the changed elements */
   modelIds?: Id64String[];
```
Comment (pmconne) [inline core/common/src/ChangedElements.ts:43]: Under what circumstances will this be undefined?

File: core/common/src/ChangedElements.ts [hunk 5]
```diff
@@ -11,26 +11,57 @@ import { AxisAlignedBox3dProps } from "./geometry/Placement";
...
+  /** Properties that changed, if any, of each changed element */
   properties?: Id64String[][];
```
Comment (pmconne) [inline core/common/src/ChangedElements.ts:45]: Does each property have a 64-bit Id? Given the class Id and property Id I can find the changed property?

File: core/common/src/ChangedElements.ts (hunk 3 above)
Comment (diegopinate) [reply core/common/src/ChangedElements.ts:30]: Yes, all arrays will have the same length, and an index relates to a single element.

File: core/common/src/ChangedElements.ts [hunk 6]
```diff
@@ -11,26 +11,57 @@ import { AxisAlignedBox3dProps } from "./geometry/Placement";
...
+  /** After state checksums of the property value
+   * Useful to determine if property values have changed
+   * between the versions being inspected
+   */
   newChecksums?: number[][];
+  /**
+   * Parent ids of the changed elements
+   * Will be "0" if the element has no parent
+   */
   parentIds?: Id64String[];
```
Comment (pmconne) [inline core/common/src/ChangedElements.ts:60]: Under what circumstances will this be undefined?

File: core/common/src/ChangedElements.ts (hunk 3 above)
Comment (pmconne) [reply core/common/src/ChangedElements.ts:30]: Please make the docs say that.

File: core/common/src/ChangedElements.ts (hunk 4 above)
Comment (diegopinate) [reply core/common/src/ChangedElements.ts:43]: If for some reason you are relying on an older agent that used to not export model Ids, same for all other nullable fields. Other fields (like properties, checksums and parent Ids) may also not be computed by an agent if it's configured to not do so for performance reasons.

File: core/common/src/ChangedElements.ts (hunk 5 above)
Comment (diegopinate) [reply core/common/src/ChangedElements.ts:45]: Oh weird, that's wrong. A property is just a string (property accessor string), will fix. And yes, with the access string and ECClass Id you can query the property value.

File: core/common/src/ChangedElements.ts (hunk 4 above)
Comment (pmconne) [reply core/common/src/ChangedElements.ts:43]: Please make the docs say that.

File: core/common/src/ChangedElements.ts (hunk 6 above)
Comment (diegopinate) [reply core/common/src/ChangedElements.ts:60]: Same as the other reason for model Ids. Will update to show that in the docs.

File: core/common/src/ChangedElements.ts [hunk 7]
```diff
@@ -11,26 +11,57 @@ import { AxisAlignedBox3dProps } from "./geometry/Placement";
  * @module Entities
```
Comment (pmconne) [inline core/common/src/ChangedElements.ts:11]: TypeDoc requires `@packageDocumentation` to appear before any non-comments, hence the docs build failure.
Put it after the copyright header, before the imports.

Comment (diegopinate) [reply core/common/src/ChangedElements.ts]: Thanks for the catch. Pushed

File: core/common/src/ChangedElements.ts (hunk 6 above)
Comment (diegopinate) [reply core/common/src/ChangedElements.ts:72]: Pushed update

File: core/common/src/ChangedElements.ts (hunk 4 above)
Comment (diegopinate) [reply core/common/src/ChangedElements.ts:50]: Pushed

File: core/common/src/ChangedElements.ts (hunk 3 above)
Comment (diegopinate) [reply core/common/src/ChangedElements.ts:30]: Pushed

File: core/common/src/ChangedElements.ts (hunk 1 above)
Comment (diegopinate) [reply core/common/src/ChangedElements.ts]: Pushed, I just got rid of it and made it "ECClass Ids" unless you think there's something that's worth pointing it to?

File: core/common/src/ChangedElements.ts (hunk 5 above)
Comment (diegopinate) [reply core/common/src/ChangedElements.ts]: Pushed fix for property. Does it make sense to show a example query here? or perhaps thats best for a tutorial/sample?

File: core/common/src/ChangedElements.ts [hunk 8]
```diff
@@ -2,35 +2,79 @@
...
+ */
 export interface ChangedElements {
+  /** Ids of elements that changed */
   elements: Id64String[];
+  /** ECClass Ids of elements that changed */
   classIds: Id64String[];
+  /** Operation that occurred on the element. Whether the element was inserted, updated or deleted.
+   * See [DbOpcode]($bentleyjs-core)
+   */
   opcodes: number[];
```
Comment (pmconne) [inline core/common/src/ChangedElements.ts:44]: If this is literally of type `DbOpcode` please change it from `number`.

File: core/common/src/ChangedElements.ts [hunk 9]
```diff
@@ -2,35 +2,79 @@
...
+  /** Type of change bitflags, see [[TypeOfChange]] */
   type: number[];
```
Comment (pmconne) [inline core/common/src/ChangedElements.ts:46]: If this is literally of type `TypeOfChange` please change it from `number`.

File: core/common/src/ChangedElements.ts [hunk 10]
```diff
@@ -2,35 +2,79 @@
...
+  /** Model Ids of the changed elements
+   * This may be undefined if the agent that did the processing job did not export model Ids
+   */
   modelIds?: Id64String[];
-  properties?: Id64String[][];
+  /** Property accessor strings of properties that changed, if any, for each changed element
+   * This may be undefined if the agent that did the processing job did not export properties
+  */
+  properties?: string[][];
```
Comment (pmconne) [inline core/common/src/ChangedElements.ts:54]: If element `x` had no changed properties is `properties[x]` an empty array, or `undefined`?

File: core/common/src/ChangedElements.ts (hunk 5 above)
Comment (pmconne) [reply core/common/src/ChangedElements.ts]: Better for tutorial/sample I think.

File: core/common/src/ChangedElements.ts (hunk 10 above)
Comment (diegopinate) [reply core/common/src/ChangedElements.ts:54]: empty array

File: core/common/src/ChangedElements.ts (hunk 9 above)
Comment (diegopinate) [reply core/common/src/ChangedElements.ts]: Pushed

File: core/common/src/ChangedElements.ts (hunk 8 above)
Comment (diegopinate) [reply core/common/src/ChangedElements.ts]: Pushed

**Review by pmconne — APPROVED** [review]

**Pull request merged**